### PR TITLE
feat: ENT-5119 Added additional fields in Algolia index for Programs

### PR DESCRIPTION
**Pull request by irfanuddinahmad**
## Description

This PR adds additional fields in the Algolia index for Programs. The additional fields are Availability, Partners, Subjects, Skills, Level Types

## Ticket Link

[ENT-5119](https://openedx.atlassian.net/browse/ENT-5119)

### CODE REVIEW

File: enterprise_catalog/apps/catalog/algolia_utils.py [hunk 1]
```diff
@@ -374,6 +376,114 @@ def get_program_type(program):
...
+
+    Arguments:
+        program (dict): a dictionary representing a program.
+
+    Returns:
+        list: a union of program courses availability.
+    """
+    # Master's programs don't have courses in the same way that our other programs do.
+    program_type = program.get('type')
+    if program_type and program_type.get('slug') == 'masters':
```
Comment (HammadAhmadWaqas) [inline enterprise_catalog/apps/catalog/algolia_utils.py]: I think `program_type` will be string, So is this  `program_type.get('slug') == 'masters'` expression right?

File: enterprise_catalog/apps/catalog/algolia_utils.py [hunk 2]
```diff
@@ -374,6 +376,114 @@ def get_program_type(program):
...
+        return _('Available now')
+
+    availability = set()
+    for course in program.get('courses', []):
+        course_status = get_course_availability(course)
+        for status in course_status:
+            availability.add(status)
+    return list(availability)
```
Comment (HammadAhmadWaqas) [inline enterprise_catalog/apps/catalog/algolia_utils.py:412]: Line 405 (return _('Available now')) returning a single item and here we are returning a list, 
Should not availability a single item always?

File: enterprise_catalog/apps/catalog/algolia_utils.py [hunk 3]
```diff
@@ -627,7 +737,14 @@ def _algolia_object_from_product(product, algolia_fields):
     elif searchable_product.get('content_type') == PROGRAM:
         searchable_product.update({
             'course_keys': get_program_course_keys(searchable_product),
+            'programs': list(get_program_type(searchable_product)),
```
Comment (HammadAhmadWaqas) [inline enterprise_catalog/apps/catalog/algolia_utils.py]: Why we are adding programs here? we have an existing key `program_type` to represent the type of the program.

File: enterprise_catalog/apps/catalog/tests/test_algolia_utils.py [hunk 4]
```diff
@@ -444,3 +451,206 @@ def test_configure_algolia_index(self, mock_search_client):
...
+                    'status': 'published',
+                    'is_enrollable': True,
+                    'is_marketable': True,
+                }]
+            }]},
+            ['Archived'],
+        ),
+    )
+    @ddt.unpack
+    def test_get_program_availability(self, program_metadata, expected_availability):
```
Comment (HammadAhmadWaqas) [inline enterprise_catalog/apps/catalog/tests/test_algolia_utils.py:568]: Are we checking availability for `Master` programs? as it is mentioned they are a little different!

File: enterprise_catalog/apps/catalog/algolia_utils.py [hunk 5]
```diff
@@ -374,6 +376,114 @@ def get_program_type(program):
...
+
+    Returns:
+        str: level type associated with the program.
+    """
+    level_types = []
+    for course in program.get('courses', []):
+        course_level_type = course.get('level_type')
+        if course_level_type:
+            level_types.append(course_level_type)
+    return max(set(level_types), key=level_types.count) if level_types else ''
```
Comment (HammadAhmadWaqas) [inline enterprise_catalog/apps/catalog/algolia_utils.py:484]: why we have used `set` here? 
can we use default param to empty value like: `max(level_types, default='', key=level_types.count)` ?

File: enterprise_catalog/apps/catalog/algolia_utils.py [hunk 6]
```diff
@@ -627,7 +737,14 @@ def _algolia_object_from_product(product, algolia_fields):
     elif searchable_product.get('content_type') == PROGRAM:
         searchable_product.update({
             'course_keys': get_program_course_keys(searchable_product),
+            'programs': list(get_program_type(searchable_product)),
+            'program_titles': list(get_program_title(searchable_product)),
```
Comment (HammadAhmadWaqas) [inline enterprise_catalog/apps/catalog/algolia_utils.py]: why we are typecasting to list? program title should not be a singular value?

**Review by HammadAhmadWaqas — CHANGES_REQUESTED** [review]

File: enterprise_catalog/apps/catalog/algolia_utils.py (hunk 1 above)
Comment (irfanuddinahmad) [reply enterprise_catalog/apps/catalog/algolia_utils.py]: Fixed.

File: enterprise_catalog/apps/catalog/algolia_utils.py (hunk 2 above)
Comment (irfanuddinahmad) [reply enterprise_catalog/apps/catalog/algolia_utils.py:412]: This is conforming to the course availability ... which is a list

File: enterprise_catalog/apps/catalog/algolia_utils.py (hunk 3 above)
Comment (irfanuddinahmad) [reply enterprise_catalog/apps/catalog/algolia_utils.py]: We want to use the same facet for courses and programs. This already exists for courses.

File: enterprise_catalog/apps/catalog/algolia_utils.py (hunk 5 above)
Comment (irfanuddinahmad) [reply enterprise_catalog/apps/catalog/algolia_utils.py:484]: We want to the find the "mode" of the list.

File: enterprise_catalog/apps/catalog/algolia_utils.py (hunk 6 above)
Comment (irfanuddinahmad) [reply enterprise_catalog/apps/catalog/algolia_utils.py]: We want to use the same field as is used for courses.

**Review by HammadAhmadWaqas — APPROVED** [review]
LGTM

**Review by muhammad-ammar — APPROVED** [review]
LGTM

**Pull request merged**